app/ui/browser-modern/views/browser/page/index.jsx

```diff
@@ -73,33 +80,75 @@ class Page extends Component {
       this.props.dispatch(UIEffects.focusURLBar(this.props.pageId, { select: isTofino }));
     });
 
+    this.webview.addEventListener('did-finish-loading', () => {
+      // Event fired only when page is successfully loaded
+      this.props.dispatch(PageActions.setPageState(this.props.pageId, {
+        load: PageState.STATES.LOADED,
+      }));
+    });
+
+    this.webview.addEventListener('did-stop-loading', () => {
+      // Event fired when a page finishes loading.
+      // May be emitted multiple times for every single frame.
+      // Emitted regardless of whether or not a failure occurs.
+      const url = this.webview.getURL();
+      const title = this.webview.getTitle();
+
+      this.props.dispatch(PageActions.setPageDetails(this.props.pageId, {
+        canGoBack: this.webview.canGoBack(),
+        canGoForward: this.webview.canGoForward(),
+        location: url,
+        title,
+      }));
+    });
+
+
     this.webview.addEventListener('did-finish-load', () => {
       // Event fired when the navigation is done and onload event is dispatched.
       // May be emitted multiple times for every single frame.
       // Not emitted when a failure occurs, unlike `did-stop-loading`.
       // Might not always be emitted after a `did-start-loading`.
     });
 
-    this.webview.addEventListener('did-fail-load', () => {
+    this.webview.addEventListener('did-fail-load', async function(e) {
       // Like 'did-finish-load', but fired when the load failed or was cancelled.
       // May be emitted multiple times for every single frame.
       // Only emitted when a failure occurs, unlike `did-stop-loading`.
-      this.props.dispatch(PageActions.setPageState(this.props.pageId, {
+      const { errorCode, errorDescription, validatedURL, isMainFrame } = e;
+
+      if (!isMainFrame) {
+        return;
+      }
+
+      // If a page is aborted (like hitting BACK/RELOAD while a page is loading), we'll get a
+      // load failure of ERR_ABORTED with code `-3` -- this is intended, so just ignore it.
+      if (Math.abs(errorCode) === 3) {
+        return;
+      }
+
+      const state = {
         load: PageState.STATES.FAILED,
-      }));
-    });
+        code: errorCode,
+        description: errorDescription,
+        url: validatedURL,
+      };
+
+      // Most cert errors are 501 from this event; the true, more descriptive
+      // error description can be retrieved from the main process. Just check
+      // here roughly if the error looks like a cert error.
+      if (Math.abs(errorCode) === 501 || /CERT/.test(errorDescription)) {
+        const { error, certificate } = await getCertificateError(validatedURL);
+        state.certificate = certificate;
+        state.description = error;
+      }
```

Comment:
All of this should be handled in a saga instead. Logic in component views should be kept at a minimum.